### PR TITLE
Matmul cleanup

### PR DESCRIPTION
**Pull request by fukatani**
Integrate matmul API by #2426 is incomplete, I clean up `_matmul` and `_batch_matmul`.

### CODE REVIEW

File: chainer/functions/math/matmul.py [hunk 1]
```diff
@@ -36,36 +34,20 @@ def _get_ld(a):
...
     if xp is numpy:
-        ret = numpy.empty(a.shape[:2] + b.shape[2:], dtype=a.dtype)
-        for i in six.moves.range(len(a)):
-            ret[i] = numpy.dot(a[i], b[i])
-        return ret
+        if a.ndim <= 2:
+            return numpy.dot(a, b)
+        else:
+            return numpy.einsum('...ij,...jk->...ik', a, b)
     return xp.matmul(a, b)
```
Comment (okuta) [inline chainer/functions/math/matmul.py]: NumPy1.10.0 has `matmul`. Do we use it?

Comment (fukatani) [inline chainer/functions/math/matmul.py]: Numpy 1.9.0 doesn't support `matmul`.
And `numpy.einsum` and `numpy.matmul` are almost equivalent performance in numpy.1.11, I don't use `numpy.matmul` now.

I implemented similar code in `depthwise_conv2d`.
https://github.com/chainer/chainer/pull/2067

If you want to use `numpy.matmul` for numpy >= 1.10, I'll replace current implementation.

Comment (fukatani) [inline chainer/functions/math/matmul.py]: `if xp is numpy and numpy.version.version.split('.')[1] == '9'` is prefer in L.46?

File: chainer/functions/math/matmul.py [hunk 2]
```diff
@@ -36,36 +34,20 @@ def _get_ld(a):
...
 
 def _matmul(a, b, transa=False, transb=False, transout=False):
-    a = array.as_mat(a)
-    b = array.as_mat(b)
     if transout:
-        # (A B)^T = B^T A^T
         transa, transb = not transb, not transa
         a, b = b, a
-    if transa:
-        a = a.T
```
Comment (okuta) [inline chainer/functions/math/matmul.py]: Which line covers this case ?

File: chainer/functions/math/matmul.py (hunk 1 above)
Comment (fukatani) [inline chainer/functions/math/matmul.py]: Sure, I fixed here.

File: chainer/functions/math/matmul.py (hunk 2 above)
Comment (fukatani) [inline chainer/functions/math/matmul.py]: if `a.ndim >= 2`, L.40 covers this case.
If `a.ndim == 1`, `a.T == a`

**Comment by fukatani** [comment]
I found this PR should include `basic_math.matmul` modification.
In #2426, we made `Matmul` behavior same as `numpy.matmul`, but `basic_math.MatmulConstVar` and `basic_math.MatmulVarConst` was not changed.

For this purpose, I added some commits.
Sorry for inconvenience.

File: chainer/functions/math/matmul.py (hunk 2 above)
Comment (okuta) [inline chainer/functions/math/matmul.py]: OK, I see.

**Comment by okuta** [comment]
Thank you!
LGTM!

**Pull request merged**